server/aesdsocket.c

```diff
@@ -14,36 +14,252 @@
 #include <stdbool.h>
 #include <syslog.h>
 #include <signal.h>
+#include "queue.h"
+#include <time.h>
+#include "aesdsocket.h"
 
 #define TCP_PORT "9000" /* Port for socket cnnection */
 #define BACKLOG 5     /* Number of pending connections in queue */
 #define FILE_PATH "/var/tmp/aesdsocketdata" /* File to write data to and send data from */
 #define BUFFERSIZE 500 /* Buffer for storing received data from client */
+#define TIMESTAMP_DUR 10 /* Period of writing timestamp (seconds) */
 
-bool signal_caught = false;
+static volatile bool signal_caught = false;
+int totalBytes = 0;
 
 /* Signal handler for SIGINT and SIGTERM,  save a copy of errno to restore later. */
-static void signal_handler(int signal_number){
-	int errno_saved = errno;
-	signal_caught = true;
-	syslog(LOG_DEBUG, "Caught signal, exiting");
-    	errno = errno_saved;
+static void signalHandler(int signal_number){
+	if (signal_number == SIGINT || signal_number == SIGTERM) {
+		int errno_saved = errno;
+		signal_caught = true;
+		syslog(LOG_DEBUG, "Caught signal, exiting");
+	    	errno = errno_saved;
+    	}
+}
+
+/* Thread function to receive data from client, write to file and send back file content to client */
+void* threadFunc(void* thread_func_args)
+{		
+	/* Cast function arguments as thread params */
+	struct connThread* thread_param = (struct connThread*) thread_func_args;
+	
+	/* Get the ip address of the connecting client */
+	char ipv4Addr[INET_ADDRSTRLEN];
+	struct sockaddr *incomingSockAddr = (struct sockaddr *)&thread_param->incomingAddr;
+    	void *binAddr = &(((struct sockaddr_in  *)incomingSockAddr)->sin_addr);
+    	const char *res = inet_ntop(thread_param->incomingAddr.ss_family, &binAddr, ipv4Addr, sizeof(ipv4Addr));
+    	if (res == NULL){
+		syslog(LOG_ERR, "Error converting client IP Address with error: %s", strerror(errno));
+		return thread_func_args;
+    	}
+	syslog(LOG_DEBUG, "Accepted connection from %s", ipv4Addr);
+	
+	/* Receive latest data packet from client into buffer */
+	int bytesRead  = 0;
+	char inBuf[BUFFERSIZE];
+	bool endStream = false;
+	int rc = 0;
+	
+	/* While data comes in through the stream */
+	while(!endStream){
+		bytesRead = recv(thread_param->connFd, inBuf, sizeof(inBuf), 0);
+		if(bytesRead <= 0){
+			endStream = true;
+			break;
+		}
+		totalBytes += bytesRead;
+		
+		/* Mutex lock file writting */
+		rc = pthread_mutex_lock(thread_param->writeMutex);
+		if ( rc != 0 ) {
+			syslog(LOG_ERR, "Mutex procurement failed with code %d\n",rc);
+			return thread_func_args;
+	    	}
+		
+		/* Append received data to file /var/tmp/aesdsocketdata */
+		int fd = open(FILE_PATH, O_RDWR | O_CREAT, S_IRWXU | S_IRWXG | S_IRWXO);
+		/* Error opening file: error printed and logged and program terminates with status 1 */
+		if (fd ==-1) {
+			syslog(LOG_ERR, "Error: (%s) while opening %s", strerror(errno), FILE_PATH);
+			return thread_func_args;
```

Comment:
@spinoandraptos what happens to `thread_param->writeMutex` in this code path and other similar ones below?